### ISSUE TITLE
Linux CI builds are failing

### ISSUE THREAD

**Issue by HighCommander4** [clangd/clangd#2244]
The latest weekly snapshot build [failed](https://github.com/clangd/clangd/actions/runs/12109291471/job/33758246137) for Linux.

The error is:

```
Post job cleanup.
/usr/bin/docker exec  60d5dd7350903409c23824b2aad1bab065e721a001a8440dc76361c23c8c1be4 sh -c "cat /etc/*release | grep ^ID"
/__e/node20/bin/node: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.28' not found (required by /__e/node20/bin/node)
```

**Comment by zyn0217** [comment]
... and so is our macOS builder. It looks like macOS 12 has been deprecated and now using it results in a hard error

> The macOS-12 environment is deprecated, consider switching to macOS-13, macOS-14 (macos-latest) or macOS-15. For more details, see https://github.com/actions/runner-images/issues/10721

**Comment by HighCommander4** [comment]
> ... and so is our macOS builder. It looks like macOS 12 has been deprecated and now using it results in a hard error
> 
> > The macOS-12 environment is deprecated, consider switching to macOS-13, macOS-14 (macos-latest) or macOS-15. For more details, see [actions/runner-images#10721](https://github.com/actions/runner-images/issues/10721)

Thanks. Filed https://github.com/clangd/clangd/issues/2245 for this.

**Comment by HighCommander4** [comment]
I guess this a regression from https://github.com/actions/runner/pull/3503. Github has removed the ability to opt into using node16 using [`ACTIONS_ALLOW_USE_UNSECURE_NODE_VERSION: true`](https://github.com/clangd/clangd/blob/909659b1ba09f5e34a57bfc9228e25ae84c67446/.github/workflows/autobuild.yaml#L9) and now forces usage of node20.

But node20 is linked against a glibc version that's newer than what's available on the OS we're trying to do the build on, which seems to be [ubuntu 18.04](https://github.com/clangd/clangd/blob/909659b1ba09f5e34a57bfc9228e25ae84c67446/.github/workflows/autobuild.yaml#L179).

Time to start building on a newer ubuntu version, I guess?

**Comment by zyn0217** [comment]
Yes. We should move to a newer image, but I'm not sure if we want to maintain the compatibility for older glibc, probably at least around 2.18 currently?

**Comment by HighCommander4** [comment]
> Yes. We should move to a newer image, but I'm not sure if we want to maintain the compatibility for older glibc, probably at least around 2.18 currently?

How did you determine that version number?

**Comment by zyn0217** [comment]
https://github.com/clangd/clangd/blob/909659b1ba09f5e34a57bfc9228e25ae84c67446/.github/workflows/autobuild.yaml#L358

**Comment by zyn0217** [comment]
From my experience, if you’re building on a system with a newer glibc, it might inadvertently pick up symbols that are missing in older versions.

We have a magic header for the compatibility purpose: 

https://github.com/clangd/clangd/blob/master/.github/workflows/lib_compat.h

So if we updated our builder system, we probably would have to add more symbols to this header as well. And this needs some investigation I think.

**Comment by HighCommander4** [comment]
I [tried a run](https://github.com/HighCommander4/clangd/actions/runs/12783039875/job/35633491493) with a ubuntu-20.04 builder, and indeed the "Check binary compatibility" step failed. The output was:

```
Run .github/workflows/lib_compat_test.py --lib=GLIBC_2.18 "$CLANGD_DIR/bin/clangd"
Binary depends on disallowed symbols above. Use some combination of:
 - relax the whitelist by adding --lib and --sym flags to this test
 - force older symbol versions by updating lib_compat.h
 - avoid dynamic dependencies by changing CMake configuration
 - remove bad dependencies from the code
exp@GLIBC_2.29
```

**Comment by zyn0217** [comment]
Thanks. It looks less thorny than I thought. Can we add that symbol to the header and see if that works?

**Comment by zyn0217** [comment]
Wait https://man7.org/linux/man-pages/man3/exp.3.html says `exp()` was introduced since 2.19 ?!

**Comment by HighCommander4** [comment]
> Can we add that symbol to the header and see if that works?

I [tried that](https://github.com/HighCommander4/clangd/commit/8a329d4f8c00ecfe30b831bcb56a588c25098b8d), but in the [new run](https://github.com/HighCommander4/clangd/actions/runs/12803993660/job/35697611492), the step still fails with the same output.

**Comment by HighCommander4** [comment]
Adjusted issue title to clarify that it's not just snapshot builds which are affected -- the official releases are built by the same Github workflow and thus are affected in the same way.

This implies we're not going to have a Linux release of clangd 20 at https://github.com/clangd/clangd/releases until this is resolved.

**Comment by InfinityWei** [comment]
> Adjusted issue title to clarify that it's not just snapshot builds which are affected -- the stable releases are built by the same Github workflow and thus are affected in the same way.
> 
> This implies we're not going to have a Linux release of clangd 20 at https://github.com/clangd/clangd/releases until this is resolved.

Maybe we can refer to this solution: https://github.com/actions/checkout/issues/1590#issuecomment-2567109195, i'm trying to test this.

**Comment by zyn0217** [comment]
I don’t think it’s still a good idea to stick with Ubuntu 18 these days, as GitHub could remove support for it at any time. @HighCommander4

**Comment by luckzylp** [comment]
I agree with that, GitHub will most likely no longer support ubuntu 18.04, on the basis of ensuring compatibility, migrating to ubuntu 20.04 is a good choice.

I [experimented the run](https://github.com/luckzylp/clangd/actions/runs/13212307259/job/36887361336) with a ubuntu-20.04 builder,  and add 
```
FORCE_SYMBOL_VERSION(exp, GLIBC_2.2.5);
```
 in lib_compat.h,  all the builder works fine, but the linux "Check binary compatibility" step is also OK. linux version of clangd is currently working fine for me in ubuntu 16.04 x64 and ubuntu 20.04 x64.

**Comment by zyn0217** [comment]
@luckzylp Thanks for exploring that. Can you put up a PR then? thanks

**Comment by luckzylp** [comment]
OK. PR: #2319

**Comment by HighCommander4** [comment]
> I [experimented the run](https://github.com/luckzylp/clangd/actions/runs/13212307259/job/36887361336) with a ubuntu-20.04 builder, and add
> 
> ```
> FORCE_SYMBOL_VERSION(exp, GLIBC_2.2.5);
> ```
> 
> in lib_compat.h, all the builder works fine, but the linux "Check binary compatibility" step is also OK. linux version of clangd is currently working fine for me in ubuntu 16.04 x64 and ubuntu 20.04 x64.

Huh. That's exactly the same thing that [I tried](https://github.com/clangd/clangd/issues/2244#issuecomment-2596826318) a few weeks ago. I wonder what has changed since then. (Somehow affected by https://github.com/clangd/clangd/pull/2306 maybe?)

**Comment by HighCommander4** [comment]
Anyways, I merged a PR and [kicked off a run](https://github.com/clangd/clangd/actions/runs/13223381475), let's see if it works.

**Comment by HighCommander4** [comment]
> Anyways, I merged the PR and [kicked off a run](https://github.com/clangd/clangd/actions/runs/13223381475), let's see if it works.

Looks like it worked: https://github.com/clangd/clangd/releases/tag/snapshot_20250208

Thanks @luckzylp and @zyn0217!
